core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -258,40 +253,46 @@ public static Object evaluateConstantExpression(
         analyzer.analyze(resolved, Scope.create());
 
         // evaluate the expression
-        return new ExpressionInterpreter(resolved, plannerContext, session, analyzer.getExpressionTypes()).evaluate();
+        return new ExpressionInterpreter(plannerContext, session).evaluate(resolved, analyzer.getExpressionTypes());
     }
 
-    public Type getType()
+    public Object evaluate(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes)
```

Comment:
> The expression is a sort of "static" program and making it constant for the interpreter allows for optimizations that span across multiple evaluations

Why would you evaluate (`optimize`) same expression multiple times? It doesn't happen in practice in codebase. However, we can optimize evaluation by caching results of evaluating different expressions.

Reply:
> resolve operators and functions and cache them, constant fold, etc

This is possible (it's not implemented yet) with `evaluate(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes)` too. Probably even more so if `ExpressionEvaluator` is used between rules.

My point is that `ExpressionEvaluator` is much more used to `optimize` same expression over and over again (with same context) by multiple rules rather than used to evaluate value of an expression with different inputs (e.g. partition pruning).

So IMO we have few options:
1. keep new interface as optimizations mentioned by you are possible with `evaluate(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes)` too. `ExpressionEvaluator` perf between rules is also greatly improved
2. move caching context outside of `ExpressionEvaluator` somehow so that it can be reused between rules
3. Have another interface, e.g. `CompiledExpression` when there would be a need for it.

My preference is 1)

Reply:
ping @martint